### DISCUSSION TITLE
🎇  Cross Section : from zero to hero! 🎇

### DISCUSSION THREAD

**Discussion by pmatran** [pymartheproject/pymarthe#19] [Show and tell]
:mega: Dear collaborators :mega:
-------------------------------------------------------------------------------------------------------

<br>

I recently started a brand new branch named `feature/modelgrid_zdim` from the main `dev`, to work on some requests suggested by @ebus-brgm few time ago.
As you already know, working on the _z-dimension_ on Marthe is kind of tricky due to the absence of clear _cell-by-cell_ definition of the layer's thickness.
The point is, the `MartheField`  objects (including the `imask`) are strickly _xy-focused_ and the _z-dimension_ is mainly _layer-based_.
In most cases, this geometric definition is good enough to work with grids and fields: we can easily visualize _xy-based_ plots of each layer, parametrize layer by layer with a given `izone`, ...
The _z-dimension_ was only required in the `vtk` utilities to extract the cell's 3D dimension. In this new feature, we've proposed to dig deeper on the _z-dimension_.

First of all, the initial `get_top_botm()` function, initially in `vtk_utils` file have been moved to the main `MartheModel` instance as a hidden method named `_get_top_bottom_arrays()`.
This method is extremly important (although unfortunately quite slow :clock430:) since it can compute vertical differences to extract the `top` and `bottom` of each _active_ cell taking into account model vertical definition: 

- `explicit` _(ex: Lizonne)_
-  `implicit` _(ex: MONA v3)_

Going further, it is now possible to extract the main `modelgrid` object with additional _z-dimension_ informations such as `zcc`, `dz`, `top`, `bottom` and `volume`.

```python
mm = MartheModel('model.rma')
mm.build_modelgrid(add_z=True)
mm.modelgrid.head()
```

Since we now have access to the vertical informations, a new required feature have been implemented: the `CrossSection` utility tool. It allows users to define, visualize and plot vertical cross sections of a `MartheModel` object. The definition of a cross section line can be done through multiple and flexible ways given a:

- path to a shapefile (unique line geometry)
- list of points coordinates forming a line
- model row/column number _(not available for nested models)_
- x/y coordinates

```python
# -- Import the cross section facility
from pymarthe.utils.xs_utils import CrossSection

# -- Cross line definition 
xs = CrossSection(mm, cross_section_line='xsline.shp') # from shapefile
xs = CrossSection(mm, cross_section_line=[(x0, y0), (x1, y1)]) # from points
xs = CrossSection(mm, cross_section_line={'row': 24}) # from row number
xs = CrossSection(mm, cross_section_line={'column': 36}) # from column number
xs = CrossSection(mm, cross_section_line={'x': 234.8}) # from x-coordinate
xs = CrossSection(mm, cross_section_line={'y': 421.2}) # from y-coodinate
```

The `CrossSection` instance was built to manage the intersected cells (nodes) by the cross section line and offer some facilites to visualize the resulting plots via internal methods:

-  `.plot()`: _xy-plot_ to localize the cross section line according to the model active domain
- `.plot_xs()`: _xz-plot_ to visualize basics informations of intersected cells vertically
- `.plot_xs_field()`: _xz-plot_ to visualize field informations of intersected cells vertically

A clear documentation have been implemented to each method and some simple examples have been added in the _*Guide_v1.0.py*_ file.
Moreover, all these functionalities have been tested successfully on _movav3_, _lizonnev2_ and _hallue_ models. :white_check_mark:

Feel free to comment, improve, suggest and test on your own models before merging this new feature to the main `dev`.
```bash
git fetch feature/modelgrid_zdim
git checkout feature/modelgrid_zdim
```

Hope you'll like it! :blush:

![crossline_xy](https://user-images.githubusercontent.com/60354180/227712102-bce41c80-9093-465c-8333-bfca0d679aaf.png)
![field_xsplot](https://user-images.githubusercontent.com/60354180/227712099-a8291326-fd7d-4f57-b6af-6db7974452f8.png)
![bylayer_add_location](https://user-images.githubusercontent.com/60354180/227712351-94eed6c5-6fbf-4fc6-994f-e6e0f0b00fc8.png)

Pierre

**Comment by ebus-brgm** [comment]
Fantastic! Thanks much Pierre for this tool!

I've made a few tests, and it works great on regular models.

However I could not make it work for a nested multi-layered model. I've tried from shapefile/points/row-col/x-y.
For the shapefile definition I get the following error:
![image](https://user-images.githubusercontent.com/82954181/228222429-6e6880ba-b137-418f-9811-d1bfb48fa171.png)
For points/row-col/x-y, I get:
![image](https://user-images.githubusercontent.com/82954181/228222580-9294f2ba-e97f-456a-b8fa-345a48115005.png)

In any case, this is awesome! and i would add another suggestion:
a function to plot the simulated water table per cell, to display on the cross-section, where we could choose the layer/the time-step of interest? something similar to the flopy method?
![image](https://user-images.githubusercontent.com/82954181/228223877-509118f4-86ef-4788-a07e-1c90206162a6.png)

best regards

**Reply by pmatran** [reply]
Hi @ebus-brgm :raising_hand:,

Thank you for your constructive feedbacks!

Let's take it slow and review each point steps-by-steps.

**1. The `MultiPolygon` issue**

This problem seems quite understable since we are performing a _unary union_ between cell bounds (taking all polygon geometries). For nested model, the larger polygons (ex: `inest=0`) are overlaying smaller polygons (ex: `inest=1`) producing a MultiPolygon entity which breaks the process. Adding a more specific node query before performing the `unary_union` will probably fix this problem **and** speed up the `.plot()` method for nested models.
However, since I do not own a multi-layered nested model myself, I couldn't perform all required tests for these kind of models. Your turn!:muscle:

**2. The `Linestring` issue**

So, what is happening over there is quite _"geo-tricky"_.:smile:
Actually, the intersection between the cross section line and the cell exterior borders is a `Point` if the line pass just over a corner of the cell or a classic `MultiPoint` on each intercepted borders . But in your case, you probably choose a vertical/horizontal cross section line that passed exactly **on** the border cell: that's the reason why the produced intersection is a `LineString` object that the pipeline cannot afford. You're quite unlucky don't you?! :sweat_smile:
But don't worry, I have already pushed a correction for that case adding a incremental translation on the cross section line position while performing the intersection with model nodes. It is based on the shortest `dx`, `dy` of the tiniest cell of the model and will absolutly not affect the results.

So, at thispoint, you will be able to carry out your previous tests on your multi-layer nested model. You can see mine below based on the model you sent me a year ago (which is single-layered).
![nested_model_plot](https://user-images.githubusercontent.com/60354180/228695525-0047fe29-1d7b-4486-a301-673ab1b6397f.png)
![nested_model_plot_xs](https://user-images.githubusercontent.com/60354180/228695555-fbc38a7b-d3c6-4c42-9051-7c7ab7cb30fd.png)
![nested_model_plot_xs_field](https://user-images.githubusercontent.com/60354180/228695567-748f5703-16a6-492f-84b5-ed7e3801a51c.png)
<br>

**3. Surfaces & Water Table feature** 

Moreover, **I totally agree** with you about this feature: it is amazing to see the water table on each cells on a cross section plot! In `flopy`, this feature is effectly quite efficient because, as all the *crosssection* part it is based on arrays and not on graphical polygon like us. In addition, the [`.plot_surface()`](https://github.com/modflowpy/flopy/blob/develop/flopy/plot/crosssection.py) method allows to plot **any type of surfaces** and not only water tables. Without any pretension, I strongly suggest to implement the same "idea" in our case. I already gave it a try in a recent `push` with the `.add_surface()` method. It seems to work well but unfortunately I only tested it for some random surface values (`top`, `bottom`, ...). Some examples below:

```python
# -- Load Lizonne model
lizonne_ws = os.path.join('examples', 'lizonnev2', 'Lizonne.rma')
mm = MartheModel(lizonne_ws)

# -- Define cross section
xs = CrossSection(mm, cross_section_line={'j': 28})

# -- Plot cross section by layer without edge lines
ax =  xs.plot_xs(lw=0, alpha=.3, vertical_ratio=8, by_layer=True)

# -- Add bottom surface on all layer (orange line)
xs.add_surface(ax, surface='bottom', color='orange')

# -- Add top surface on layer 3 (black line)
xs.add_surface(ax, surface='top', layer=3, lw=1.05, color='k')

# -- Add middle thickness on layer 11 (blue line)
surface = xs.xs_df[['top', 'bottom']].apply('mean', axis=1)
xs.add_surface(ax, surface=surface, layer=11, lw=1.1)

# -- Draw cross section
plt.show()
```
![nested_model_add_surfaces](https://user-images.githubusercontent.com/60354180/228697085-b0c622ea-5d06-4009-a48d-bd44fdbab0b3.png)
<br>

Unfortunately, I do not have a valid `chasim.out` file of any model to test it on a water table values. But, since the method can manage a `MartheField` instance as input surface, I really think it won't be difficult. I give you some hints to carry this out below:

```python
# -- Create a MartheFieldSeries instance
from pymarthe.mfield import MartheField, MartheFieldSeries
chasim = 'chasim.out'
mfs = MartheFieldSeries(mm, chasim=chasim)

# -- Load hydraulic head on a specific available timestemp
istep = 10
mfs.load_field(field='CHARGE', istep=istep)

# -- Extract the MartheField instance of the required istep
surface = mfs.data['CHARGE'][istep]

# -- Define cross section
xs = CrossSection(mm, cross_section_line='xsline.shp')

# -- Plot cross section by layer without edge lines
ax = xs.plot_xs(lw=0, alpha=.3, by_layer=True)

# -- Add hydraulic head on required layer
xs.add_surface(ax, surface=surface, layer=4, lw=1.1)

# -- Add title
ax.set_title(f'Hydraulic Head\nmodel:{mm.mlname}, istep={istep}', fontsize=10)
```

Anyway, thank's again for your comments, we will probably let this branch open for a while to have enough time to implement and test these features.:thumbsup:

Regards,

_Pierre_

**Comment by pmatran** [comment]
Hi everyone :wave:,

I just wanted to remind you that the branch supporting the new `CrossSection`  feature ([feature/modelgrid_zdim](https://github.com/apryet/adeqwat/tree/feature/modelgrid_zdim)) is currently being merged into the `dev` branch.

Feel free to:
+ :clipboard: review the code
+  :collision: Perform some tests on your models
+  :white_check_mark: Validate the **PR**   &nbsp;&nbsp;&nbsp; :arrow_right: #20

<br>

Please do not hesitate to contact me for further details if necessary (:grey_question:)

<br>

Regards,

_Pierre_

**Reply by ebus-brgm** [reply]
Hello Pierre, I've been testing this new feature recently and it is awesome. Thanks much for this contribution!
However i am encoutering a strange issue using the .plot_xs_field() function
I am using a single-layer, regular grid model. I am plotting the cross section by defining coordinates, e.g.:
xs = CrossSection(mm, cross_section_line=[(292, 340), (292,400) ])
I am plotting the chasim.out ("%SATURAT" parameter), using both plot_xs_field() and add_surface()
i am also plotting the water table ("CHARGE") using add_surface()

When I am plotting a cross section, the plot_xs_field() does not provide the right value per cell. As illustrated below:
I plotted the saturation field (0-100%) and the saturation surface (thus displaying between 0 and 100 on the vertical axis) so that the two can be compared (and should be identical). 
we should see some red in the cells (0% sat) where the simulated water table is below or close to the bottom; but it is not the case in the cross-section. However the same field displayed using the .add_surface() function is correct  (red line)
![image](https://github.com/apryet/adeqwat/assets/82954181/deeba304-0114-4311-90d0-811fbcdab7bd)

I should take the time to look into the code...but if it rings any bells... Many thanks!!

**Reply by pmatran** [reply]
<br>

Hi @ebus-brgm :wave:
-------------------------------------------------------------------------

I'm glad that you like this feature and find it useful :pray:. I don't know if you made this graph with the `CrossSection` instance, but it's awesome :star_struck: :star_struck: :star_struck:

First of all, I'd like to thank you very much for your feedback and for the quality of your request. We understand the problem straight away 'cause you've explained the context and the different functions you're targeting.

There does seem to be a problem with the values in the `%SATURATION` field. I'd say it's a problem with assigning field values to the matplotlib `patches`. On the other hand, the values on the graph don't seem to be completely _random_, there's some form of continuity between values, in other words, when a cell value along the cross section line is `~90%`, the neighbouring nodes are not so far from this value, that would not seem very likely (without any knownledge about the the precision of the mesh).
On the face of it, I may have a small idea on the question :bulb:, but as I don't have access to the model, I'm going to  ask you to carry out the test by yourself :construction_worker_man:.

So, would it be possible for you to replace in the file `pymarthe.utils.xs_utils` (_l.290_):

```python
xsnodes = (
		self.modelgrid
		.loc[
			np.tile(
				(
                                        self.modelgrid
				        .query('layer == 0')['vertices']
				        .map(Polygon)
				        .map(self.xsline.intersects)
                                ),
				reps=self.mm.nlay
			)
		]
		.index
		.values
	)
```

By the following code:

```python
xsnodes = (
		self.modelgrid
		.loc[
			np.tile(
				(
					self.modelgrid
					.query('layer == 0')['vertices']
					.map(Polygon)
					.apply(lambda p: 
							( 
							translate(self.xsline,
									xoff=self.offset,
									yoff=self.offset)
							.intersects(p)
							)
						)
				),
				reps=self.mm.nlay
			)
		]
		.index
		.values
	)
```

If the problem persists, be sure to check that you haven't made any mistakes in the fields your using and I suggest you to create a new issue, explaining the context and attaching a _zip_ file :open_file_folder: containing the main model (anonymised if necessary) and the **python script** used so that all your collaborators :busts_in_silhouette: can access and work on the debugging.

Keep me informed and thanks again,

@pmatran

**Reply by ebus-brgm** [reply]
Hello @pmatran! thanks much for your answer and effort. Yes this graph was made with pymarthe!
I've tried the correction you suggested and it does seem to solve the problem !! Many thanks! 
There is a small drawback, the _apply()_ function makes it much much slower, than the original _.map(self.xsline.intersects)_... (i.e. it takes a few minutes instead of a few seconds). But it is correct!
![image](https://github.com/apryet/adeqwat/assets/82954181/65ee8c3c-a982-48c6-9f82-7a779117f208)


Cheers!

**Reply by pmatran** [reply]
I knew it! What a shot @ebus-brgm  :ok_hand: 

The fact that it's much slower than before  not comes from the use of `.apply()` instead of `.map()` , it's rather the translation of the cross section that is executed for each polygon instead of being created beforehand and stored in a protected variable `._xsline` (that was just a test). 

<br>

Anyway, would you mind creating an issue related to your previous message (just click _"Reference in new issue"_) please :pray: ? I will try to fix  better that problem (when I have spare time) conserving run time efficiency :wink:

Cheers!
